src/control/server/storage/scm/provider.go

```diff
@@ -789,3 +816,80 @@ func (p *Provider) unmount(target string, flags int) (*MountResponse, error) {
 func (p *Provider) IsMounted(target string) (bool, error) {
 	return p.sys.IsMounted(target)
 }
+
+func (p *Provider) getUIDs(requested []string) ([]string, error) {
+	modules, err := p.backend.Discover()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(modules) == 0 {
+		return nil, errors.New("no SCM modules")
+	}
+
+	uids := make([]string, 0, len(modules))
+	if len(requested) == 0 {
+		for _, m := range modules {
+			uids = append(uids, m.UID)
+		}
+	} else {
+		uids = requested
+	}
+
+	return uids, nil
+}
+
+// QueryFirmware fetches the status of SCM device firmware.
+func (p *Provider) QueryFirmware(req FirmwareQueryRequest) (*FirmwareQueryResponse, error) {
+	if p.shouldForward(req) {
+		return p.fwFwd.Query(req)
+	}
+
+	uids, err := p.getUIDs(req.Devices)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &FirmwareQueryResponse{
+		FirmwareInfo: make(map[string]storage.ScmFirmwareInfo, len(uids)),
+	}
+	for _, uid := range uids {
+		fwInfo, err := p.backend.GetFirmwareStatus(uid)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error getting firmware status for device %s", uid)
+		}
+		resp.FirmwareInfo[uid] = *fwInfo
+	}
+
+	return resp, nil
+}
+
+// UpdateFirmware updates the SCM device firmware.
+func (p *Provider) UpdateFirmware(req FirmwareUpdateRequest) (*FirmwareUpdateResponse, error) {
+	if p.shouldForward(req) {
+		return p.fwFwd.Update(req)
+	}
+
+	if len(req.FirmwarePath) == 0 {
+		return nil, errors.New("missing path to firmware file")
+	}
```

Comment:
I guess validating that the path resolves to a real file would make testing a bit more complicated. Do we know what happens when a bogus path is given to the update code?

Reply:
We actually check the existence of the file in our ipmctl bindings package before sending it down to libipmctl, and return an error if the file doesn't exist. I did that because the error returned from libipmctl is sort of unhelpful. I think it was something like "file dump failed."